extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/caffeine/CaffeineCacheImpl.java

```diff
@@ -135,7 +135,9 @@ private <K, V> CompletableFuture<Object> getFromCaffeine(K key, Function<K, V> v
             statsCounter.recordMisses(1);
             try {
                 Object value = valueLoader.apply(key);
-                newCacheValue.complete(NullValueConverter.toCacheValue(value));
+                if (value == null && cacheInfo.cacheNullValues) {
+                    newCacheValue.complete(NullValueConverter.toCacheValue(value));
+                }
```

Comment:
At this point of the method execution, we've already put `newCacheValue` into the Caffeine cache. It hasn't been completed yet, is  returned by the method and used to build a `Uni` which will never emit any item.

This is one step closer to the required changes:
```suggestion
                if (value != null || cacheInfo.cacheNullValues) {
                    newCacheValue.complete(NullValueConverter.toCacheValue(value));
                } else {
                    cache.asMap().remove(key, newCacheValue);
                    newCacheValue.complete(null);
                    return newCacheValue;
                }
```
Concurrent calls to the cached method will be an issue though. If it is called from thread A (first) and thread B (second) concurrently, thread B may retrieve the uncompleted `newCacheValue` from the cache while thread A is processing the value. Ideally, thread B should process its own value and never retrieve anything that resulted from thread A execution. I need to think a bit more about that.